File: web/NextJS/src/lib/sets/setMinimumOccupancy.test.ts
import { describe, expect, it } from "vitest";

import { API_ERROR_CODES, ApiError } from "@/lib/api/errors";
import {
  assertSetMinimumOccupancy,
  countModPieces,
  countWeaponOrArmorItems,
  evaluateSetMinimumOccupancy,
} from "@/lib/sets/setMinimumOccupancy";

describe("countWeaponOrArmorItems", () => {
  it("counts unique weapon domain slots", () => {
    expect(
      countWeaponOrArmorItems("weapon", [
        { slot: "primary" },
        { slot: "heavy" },
        { slot: "primary" },
      ]),
    ).toBe(2);
  });

  it("ignores soft-removed", () => {
    expect(
      countWeaponOrArmorItems("armor", [
        { slot: "helmet" },
        { slot: "arms", removedAt: "2026-01-01" },
      ]),
    ).toBe(1);
  });
});

describe("countModPieces", () => {
  it("counts distinct armor pieces from mod keys", () => {
    expect(
      countModPieces([
        { slot: "helmet:1" },
        { slot: "helmet:2" },
        { slot: "arms:3" },
      ]),
    ).toBe(2);
  });

  it("treats legacy free-list as one piece", () => {
    expect(
      countModPieces([{ slot: "mod:10" }, { slot: "mod:11" }]),
    ).toBe(1);
  });
});

describe("evaluateSetMinimumOccupancy", () => {
  it("allows empty weapon scaffold", () => {
    const r = evaluateSetMinimumOccupancy("weapon", []);
    expect(r.ok).toBe(true);
    if (r.ok) expect(r.empty).toBe(true);
  });

  it("rejects single-item weapon set", () => {
    const r = evaluateSetMinimumOccupancy("weapon", [{ slot: "primary" }]);
    expect(r.ok).toBe(false);
    if (!r.ok) {
      expect(r.code).toBe(API_ERROR_CODES.SET_MIN_ITEMS);
      expect(r.count).toBe(1);
      expect(r.required).toBe(2);
    }
  });

  it("accepts weapon set with two items", () => {
    const r = evaluateSetMinimumOccupancy("weapon", [
      { slot: "primary" },
      { slot: "special" },
    ]);
    expect(r.ok).toBe(true);
    if (r.ok) expect(r.empty).toBe(false);
  });

  it("rejects armor set with one piece", () => {
    const r = evaluateSetMinimumOccupancy("armor", [{ slot: "helmet" }]);
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.code).toBe(API_ERROR_CODES.SET_MIN_ITEMS);
  });

  it("rejects mod set with one piece only", () => {
    const r = evaluateSetMinimumOccupancy("mod", [
      { slot: "helmet:1" },
      { slot: "helmet:2" },
    ]);
    expect(r.ok).toBe(false);
    if (!r.ok) {
      expect(r.code).toBe(API_ERROR_CODES.MOD_SET_MIN_SLOTS);
      expect(r.count).toBe(1);
    }
  });

  it("accepts mod set with two pieces", () => {
    const r = evaluateSetMinimumOccupancy("mod", [
      { slot: "helmet:1" },
      { slot: "legs:2" },
    ]);
    expect(r.ok).toBe(true);
  });

  it("rejects incomplete pair", () => {
    const r = evaluateSetMinimumOccupancy("pair", [
      { slot: "exotic_weapon" },
    ]);
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.code).toBe(API_ERROR_CODES.PAIR_INCOMPLETE);
  });

  it("accepts complete pair", () => {
    const r = evaluateSetMinimumOccupancy("pair", [
      { slot: "exotic_weapon" },
      { slot: "exotic_armor" },
    ]);
    expect(r.ok).toBe(true);
  });

  it("always ok for fashion", () => {
    expect(evaluateSetMinimumOccupancy("fashion", []).ok).toBe(true);
    expect(
      evaluateSetMinimumOccupancy("fashion", [{ slot: "ghost" }]).ok,
    ).toBe(true);
  });
});

describe("assertSetMinimumOccupancy", () => {
  it("throws ApiError with SET_MIN_ITEMS", () => {
    try {
      assertSetMinimumOccupancy("weapon", [{ slot: "heavy" }]);
      expect.unreachable();
    } catch (e) {
      expect(e).toBeInstanceOf(ApiError);
      expect((e as ApiError).code).toBe(API_ERROR_CODES.SET_MIN_ITEMS);
    }
  });

  it("does not throw for empty scaffold", () => {
    expect(() => assertSetMinimumOccupancy("mod", [])).not.toThrow();
  });
});

File: flutter/packages/domain/test/set_minimum_occupancy_test.dart
import 'package:destiny2_domain/destiny2_domain.dart';
import 'package:test/test.dart';

/// DAC-DST-010 / DAC-DST-011 / DAC-SET-002 package minimum occupancy.
void main() {
  group('countWeaponOrArmorItems', () {
    test('counts unique weapon domain slots', () {
      expect(
        countWeaponOrArmorItems(SetType.weapon, const [
          SetOccupancyItem(slot: 'primary'),
          SetOccupancyItem(slot: 'heavy'),
          SetOccupancyItem(slot: 'primary'),
        ]),
        2,
      );
    });

    test('ignores soft-removed', () {
      expect(
        countWeaponOrArmorItems(SetType.armor, const [
          SetOccupancyItem(slot: 'helmet'),
          SetOccupancyItem(slot: 'arms', removedAt: '2026-01-01'),
        ]),
        1,
      );
    });
  });

  group('countModPieces', () {
    test('counts distinct armor pieces from mod keys', () {
      expect(
        countModPieces(const [
          SetOccupancyItem(slot: 'helmet:1'),
          SetOccupancyItem(slot: 'helmet:2'),
          SetOccupancyItem(slot: 'arms:3'),
        ]),
        2,
      );
    });

    test('treats legacy free-list as one piece', () {
      expect(
        countModPieces(const [
          SetOccupancyItem(slot: 'mod:10'),
          SetOccupancyItem(slot: 'mod:11'),
        ]),
        1,
      );
    });
  });

  group('evaluateSetMinimumOccupancy — DAC-DST-010 weapon/armor', () {
    test('rejects empty weapon scaffold with SET_MIN_ITEMS', () {
      final r = evaluateSetMinimumOccupancy(SetType.weapon, const []);
      expect(r.ok, isFalse);
      expect(r.empty, isTrue);
      expect(r.code, DomainFailureCodes.setMinItems);
      expect(r.count, 0);
      expect(r.required, 2);
    });

    test('rejects single-item weapon set', () {
      final r = evaluateSetMinimumOccupancy(SetType.weapon, const [
        SetOccupancyItem(slot: 'primary'),
      ]);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.setMinItems);
      expect(r.count, 1);
      expect(r.required, 2);
      expect(r.message, contains('Weapon'));
    });

    test('accepts weapon set with two items', () {
      final r = evaluateSetMinimumOccupancy(SetType.weapon, const [
        SetOccupancyItem(slot: 'primary'),
        SetOccupancyItem(slot: 'special'),
      ]);
      expect(r.ok, isTrue);
      expect(r.empty, isFalse);
      expect(r.count, 2);
    });

    test('rejects armor set with one piece', () {
      final r = evaluateSetMinimumOccupancy(SetType.armor, const [
        SetOccupancyItem(slot: 'helmet'),
      ]);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.setMinItems);
      expect(r.message, contains('Armor'));
    });

    test('accepts armor set with two pieces', () {
      final r = evaluateSetMinimumOccupancy(SetType.armor, const [
        SetOccupancyItem(slot: 'helmet'),
        SetOccupancyItem(slot: 'legs'),
      ]);
      expect(r.ok, isTrue);
    });
  });

  group('evaluateSetMinimumOccupancy — DAC-DST-011 mod multi-piece', () {
    test('rejects empty mod set with MOD_SET_MIN_SLOTS', () {
      final r = evaluateSetMinimumOccupancy(SetType.mod, const []);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.modSetMinSlots);
      expect(r.count, 0);
    });

    test('rejects mod set with one piece only (multiple plugs same piece)', () {
      final r = evaluateSetMinimumOccupancy(SetType.mod, const [
        SetOccupancyItem(slot: 'helmet:1'),
        SetOccupancyItem(slot: 'helmet:2'),
      ]);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.modSetMinSlots);
      expect(r.count, 1);
    });

    test('accepts mod set with two pieces', () {
      final r = evaluateSetMinimumOccupancy(SetType.mod, const [
        SetOccupancyItem(slot: 'helmet:1'),
        SetOccupancyItem(slot: 'legs:2'),
      ]);
      expect(r.ok, isTrue);
      expect(r.count, 2);
    });
  });

  group('evaluateSetMinimumOccupancy — DAC-SET-002 pair complete', () {
    test('rejects incomplete pair (weapon only)', () {
      final r = evaluateSetMinimumOccupancy(SetType.pair, const [
        SetOccupancyItem(slot: 'exotic_weapon'),
      ]);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.pairIncomplete);
      expect(r.count, 1);
    });

    test('rejects incomplete pair (armor only)', () {
      final r = evaluateSetMinimumOccupancy(SetType.pair, const [
        SetOccupancyItem(slot: 'exotic_armor'),
      ]);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.pairIncomplete);
    });

    test('rejects empty pair', () {
      final r = evaluateSetMinimumOccupancy(SetType.pair, const []);
      expect(r.ok, isFalse);
      expect(r.code, DomainFailureCodes.pairIncomplete);
      expect(r.empty, isTrue);
    });

    test('accepts complete pair', () {
      final r = evaluateSetMinimumOccupancy(SetType.pair, const [
        SetOccupancyItem(slot: 'exotic_weapon'),
        SetOccupancyItem(slot: 'exotic_armor'),
      ]);
      expect(r.ok, isTrue);
      expect(r.count, 2);
    });
  });

  group('evaluateSetMinimumOccupancy — fashion exempt', () {
    test('always ok for fashion', () {
      expect(evaluateSetMinimumOccupancy(SetType.fashion, const []).ok, isTrue);
      expect(
        evaluateSetMinimumOccupancy(SetType.fashion, const [
          SetOccupancyItem(slot: 'ghost'),
        ]).ok,
        isTrue,
      );
    });
  });

  group('setWouldPassSaveRules', () {
    test('false under min, true at min, fashion always true', () {
      expect(
        setWouldPassSaveRules(SetType.weapon, const [
          SetOccupancyItem(slot: 'primary'),
        ]),
        isFalse,
      );
      expect(
        setWouldPassSaveRules(SetType.weapon, const [
          SetOccupancyItem(slot: 'primary'),
          SetOccupancyItem(slot: 'heavy'),
        ]),
        isTrue,
      );
      expect(setWouldPassSaveRules(SetType.fashion, const []), isTrue);
    });
  });

  group('formatSetOccupancyMessage', () {
    test('plain language for SET_MIN_ITEMS', () {
      final msg = formatSetOccupancyMessage(
        code: DomainFailureCodes.setMinItems,
        setType: SetType.weapon,
        count: 1,
        required: 2,
      );
      expect(msg, isNot(contains(RegExp(r'^[0-9a-f]{8,}$'))));
      expect(msg.toLowerCase(), contains('weapon'));
      expect(msg, contains('2'));
    });
  });
}
